plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -0,0 +1,123 @@
+
+let f = \x.\y -> x*y
+    y = f q
+in f a b + y
+
+Here we have a `f` that is not fully applied via `y`, so we cannot eliminate the let-binding of `f`.
+TODO
+To do this we need to count the number of type/term lambda abstractions,
+so we know how many term/type arguments they have.
+We also need to decide whether or not to inline based on the call site, which we don’t currently.
+
+The heuristic: Counting lambdas and arguments
+
+We check for a binding that is a lambda abstraction and is fully applied in the body.
+
+E.g.,
+
+
+We pattern match the _rhs_ with `LamAbs` or `TyAbs` (lambda abstraction for terms or types),
+and count the number of lambdas.
+Then, in the _body_, we check for term or type application (`Apply` or `TyInst`) of _v_.
+
+If _v_ is fully applied in the body, i.e., if
+
+1. the number of type lambdas equals the number of type arguments applied, and
+2. the number of term lambdas equals the number of term arguments applied, and
+
+if other call site inlining conditions are satisfied,
+  (we currently inline too little, this will be improved later, see below)
+
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_.
+For the rest of the discussion here we focus on the conditions 1 and 2 only.
+
+E.g., for
+
+```haskell
+let f = \x.\y.x+y in
+    f 2 4
+```
+
+the _body_ `f 2 4` is a function application of `f`. Since `f` is fully applied, we inline `f`.
+
+### Examples
+
+Below are some examples that involve a _body_ that is not fully reducible/applied but
+following our heuristic is beneficial.
+
+#### Example 1: function in body
+
+```haskell
+let f = \x . x
+     g = \y . f
+in g a
+```
+
+`f` and `g` each has 1 lambda. However, `g`'s _body_ includes `f` which also has a lambda.
+Since we only count the number of lambdas, `g` is fully applied, and we inline.
+`g a` reduces to `f`, which reduces the amount of code even though `f` has a lambda that
+is not applied.
+
+#### Example 2: function as an argument
+
+```haskell
+let id :: forall a . a -> a
+    id x = x
+in id {int -> int} (\x -> x +1)
+```
+
+Here we have a type application for a function that takes one type variable.
+I.e., it's fully applied in terms of type.
+In terms of terms, `id` takes one argument, and is indeed applied to one.
+So `id` is fully applied! And we inline it.
+
+Reducing the _body_ `id {int -> int} (\x -> x +1)` leaves us `(\x -> x +1)`,
+a function with 1 lambda. Even though it's not fully applied, inlining
+and reducing `id` reduces the amount of code, as desired.
+
+#### Example 3: function application in _RHS_
+
+```haskell
+let f = (\x.\y.x+y) 4
+in f 5
+```
+
+With beta-reduction, `f` becomes `\y.4+y` and it has 1 lambda.
+The _body_ `f 5` is a fully applied function!
+We can reduce it to 9.
+However, because the _rhs_ `(\x.\y.x+y) 4` is a function application, not a lambda abstraction,
+we won't inline it with the current implementation. This is something we may be able to improve.
```

Comment:
We will inline it if we run beta-reduction first, before the inliner. Even if we run inliner first, as long as the simplifier runs two or more passes, it will still get inlined.